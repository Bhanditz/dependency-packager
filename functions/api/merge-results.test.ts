import { ILambdaResponse } from "./";
import mergeResults from "./merge-results";

describe("mergeResults", () => {
  const react: ILambdaResponse = {
    contents: {
      "/node_modules/react/lib/react.development.js": "yes",
      "/node_modules/fbjs/lib/index.js": "yes yes",
    },
    dependency: {
      name: "react",
      version: "16.0.0",
    },
    dependencyDependencies: {
      fbjs: {
        entries: ["fbjs"],
        parents: ["react"],
        resolved: "12.0.0",
        semver: "^12.0.0",
      },
    },
    peerDependencies: {},
  };
  const reactDom: ILambdaResponse = {
    contents: {
      "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
      "/node_modules/fbjs/lib/index.js": "yes yes",
    },
    dependency: {
      name: "react-dom",
      version: "16.0.0",
    },
    dependencyDependencies: {
      fbjs: {
        entries: ["fbjs"],
        parents: ["react-dom"],
        resolved: "12.0.0",
        semver: "^12.0.0",
      },
    },
    peerDependencies: {},
  };
  it("can merge 2 responses", () => {
    const merge = mergeResults([react, reactDom]);

    expect(merge).toMatchSnapshot();
  });

  it("can merge responses with absolute conflicts", () => {
    const conflict: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes",
      },
      dependency: {
        name: "conflict",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs"],
          parents: ["conflict"],
          resolved: "12.0.1",
          semver: "^12.0.0",
        },
      },
      peerDependencies: {},
    };

    const merge = mergeResults([react, reactDom, conflict]);

    expect(merge).toMatchSnapshot();
  });

  it("can merge responses with absolute+semver conflicts, but same entries", () => {
    const conflict: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes",
      },
      dependency: {
        name: "conflict",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs"],
          parents: ["conflict"],
          resolved: "12.0.1",
          semver: "^12.2.0",
        },
      },
      peerDependencies: {},
    };

    const merge = mergeResults([react, reactDom, conflict]);

    expect(merge).toMatchSnapshot();
  });

  it("can merge responses with full semver conflicts", () => {
    const conflict: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes 11",
      },
      dependency: {
        name: "conflict",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs"],
          parents: ["conflict"],
          resolved: "11.0.1",
          semver: "^11.2.0",
        },
      },
      peerDependencies: {},
    };

    const merge = mergeResults([react, reactDom, conflict]);

    expect(merge).toMatchSnapshot();
  });

  it("can merge responses with semver/absolute conflicts, but different entries", () => {
    const conflict: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes c",
      },
      dependency: {
        name: "conflict",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs", "fbjs/lib/test.js"],
          parents: ["conflict"],
          resolved: "12.0.1",
          semver: "^12.2.0",
        },
      },
      peerDependencies: {},
    };

    const merge = mergeResults([react, reactDom, conflict]);

    expect(merge).toMatchSnapshot();
  });

  it("can merge responses with conflicts recursively", () => {
    const conflict1: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes c",
      },
      dependency: {
        name: "conflict1",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs", "fbjs/lib/test.js"],
          parents: ["conflict1"],
          resolved: "12.0.1",
          semver: "^12.2.0",
        },
      },
      peerDependencies: {},
    };

    const conflict2: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes 11.0.2",
      },
      dependency: {
        name: "conflict2",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs", "fbjs/lib/test.js"],
          parents: ["conflict2"],
          resolved: "11.0.2",
          semver: "^11.2.0",
        },
      },
      peerDependencies: {},
    };

    const merge = mergeResults([react, reactDom, conflict1, conflict2]);

    expect(merge).toMatchSnapshot();
  });

  it("can merge responses with multiple conflicts", () => {
    const conflict1: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes c",
      },
      dependency: {
        name: "conflict1",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs", "fbjs/lib/test.js"],
          parents: ["conflict1"],
          resolved: "12.0.1",
          semver: "^12.2.0",
        },
      },
      peerDependencies: {},
    };

    const conflict2: ILambdaResponse = {
      contents: {
        "/node_modules/react/lib/react.development.js": "yes",
        "/node_modules/fbjs/lib/index.js": "yes yes",
      },
      dependency: {
        name: "conflict2",
        version: "16.0.0",
      },
      dependencyDependencies: {
        fbjs: {
          entries: ["fbjs", "fbjs/lib/test.js"],
          parents: ["conflict2"],
          resolved: "12.0.1",
          semver: "^12.2.0",
        },
      },
      peerDependencies: {},
    };

    const merge = mergeResults([react, reactDom, conflict1, conflict2]);

    expect(merge).toMatchSnapshot();
  });
});
