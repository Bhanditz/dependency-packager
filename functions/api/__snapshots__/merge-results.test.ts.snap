// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeResults can merge 2 responses 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/lib/index.js": "yes yes",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {},
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;

exports[`mergeResults can merge responses with absolute conflicts 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/12.0.0/lib/index.js": "yes yes",
    "/node_modules/fbjs/lib/index.js": "yes yes",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "conflict",
      "version": "16.0.0",
    },
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {
    "react": Object {
      "fbjs": "fbjs/12.0.0",
    },
    "react-dom": Object {
      "fbjs": "fbjs/12.0.0",
    },
  },
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "conflict",
      ],
      "resolved": "12.0.1",
      "semver": "^12.0.0",
    },
    "fbjs/12.0.0": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;

exports[`mergeResults can merge responses with absolute+semver conflicts, but same entries 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/12.0.0/lib/index.js": "yes yes",
    "/node_modules/fbjs/lib/index.js": "yes yes",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "conflict",
      "version": "16.0.0",
    },
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {
    "react": Object {
      "fbjs": "fbjs/12.0.0",
    },
    "react-dom": Object {
      "fbjs": "fbjs/12.0.0",
    },
  },
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "conflict",
      ],
      "resolved": "12.0.1",
      "semver": "^12.2.0",
    },
    "fbjs/12.0.0": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;

exports[`mergeResults can merge responses with conflicts recursively 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/11.0.2/lib/index.js": "yes yes 11.0.2",
    "/node_modules/fbjs/12.0.0/lib/index.js": "yes yes",
    "/node_modules/fbjs/lib/index.js": "yes yes c",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "conflict1",
      "version": "16.0.0",
    },
    Object {
      "name": "conflict2",
      "version": "16.0.0",
    },
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {
    "conflict2": Object {
      "fbjs": "fbjs/11.0.2",
    },
    "react": Object {
      "fbjs": "fbjs/12.0.0",
    },
    "react-dom": Object {
      "fbjs": "fbjs/12.0.0",
    },
  },
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
        "fbjs/lib/test.js",
      ],
      "parents": Array [
        "conflict1",
      ],
      "resolved": "12.0.1",
      "semver": "^12.2.0",
    },
    "fbjs/11.0.2": Object {
      "entries": Array [
        "fbjs",
        "fbjs/lib/test.js",
      ],
      "parents": Array [
        "conflict2",
      ],
      "resolved": "11.0.2",
      "semver": "^11.2.0",
    },
    "fbjs/12.0.0": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;

exports[`mergeResults can merge responses with full semver conflicts 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/12.0.0/lib/index.js": "yes yes",
    "/node_modules/fbjs/lib/index.js": "yes yes 11",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "conflict",
      "version": "16.0.0",
    },
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {
    "react": Object {
      "fbjs": "fbjs/12.0.0",
    },
    "react-dom": Object {
      "fbjs": "fbjs/12.0.0",
    },
  },
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "conflict",
      ],
      "resolved": "11.0.1",
      "semver": "^11.2.0",
    },
    "fbjs/12.0.0": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;

exports[`mergeResults can merge responses with multiple conflicts 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/12.0.0/lib/index.js": "yes yes",
    "/node_modules/fbjs/lib/index.js": "yes yes",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "conflict1",
      "version": "16.0.0",
    },
    Object {
      "name": "conflict2",
      "version": "16.0.0",
    },
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {
    "react": Object {
      "fbjs": "fbjs/12.0.0",
    },
    "react-dom": Object {
      "fbjs": "fbjs/12.0.0",
    },
  },
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
        "fbjs/lib/test.js",
      ],
      "parents": Array [
        "conflict1",
        "conflict2",
      ],
      "resolved": "12.0.1",
      "semver": "^12.2.0",
    },
    "fbjs/12.0.0": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;

exports[`mergeResults can merge responses with semver/absolute conflicts, but different entries 1`] = `
Object {
  "contents": Object {
    "/node_modules/fbjs/12.0.0/lib/index.js": "yes yes",
    "/node_modules/fbjs/lib/index.js": "yes yes c",
    "/node_modules/react-dom/lib/react-dom.development.js": "yes yes",
    "/node_modules/react/lib/react.development.js": "yes",
  },
  "dependencies": Array [
    Object {
      "name": "conflict",
      "version": "16.0.0",
    },
    Object {
      "name": "react",
      "version": "16.0.0",
    },
    Object {
      "name": "react-dom",
      "version": "16.0.0",
    },
  ],
  "dependencyAliases": Object {
    "react": Object {
      "fbjs": "fbjs/12.0.0",
    },
    "react-dom": Object {
      "fbjs": "fbjs/12.0.0",
    },
  },
  "dependencyDependencies": Object {
    "fbjs": Object {
      "entries": Array [
        "fbjs",
        "fbjs/lib/test.js",
      ],
      "parents": Array [
        "conflict",
      ],
      "resolved": "12.0.1",
      "semver": "^12.2.0",
    },
    "fbjs/12.0.0": Object {
      "entries": Array [
        "fbjs",
      ],
      "parents": Array [
        "react",
        "react-dom",
      ],
      "resolved": "12.0.0",
      "semver": "^12.0.0",
    },
  },
}
`;
